File: c++/maxFlow1/main.cpp
#include <bits/stdc++.h>

using namespace std;
vector<long long> G[205];

vector<int> visited;

vector<int> dfs(int node) {
    visited.push_back(node);
    if (node == 201) {
        vector<int> p;
        p.push_back(201);
        return p;
    } else
        for (int neighbor : G[node]) {
            bool isContained = false;
            for (int elem : visited)
                if (elem == neighbor)
                    isContained = true;
            if (!isContained) {
                vector<int> neighborPath = dfs(neighbor);
                if (!neighborPath.empty()) {
                    neighborPath.insert(neighborPath.begin(), node);
                    return neighborPath;
                }
            }
        }
    return vector<int>();
}

int maxFlow() {
    visited.clear();
    vector<int> path = dfs(0);
    if (path.empty())
        return 0;
    for (int i = 0; i < path.size() - 1; i++) {
        int a = path[i];
        int b = path[i + 1];
        for (auto iter = G[a].begin(); iter != G[a].end(); iter++) {
            if (*iter == b) {
                G[a].erase(iter);
                break;
            }
        }
        G[b].push_back(a);
    }
    return maxFlow() + 1;
}

int main() {
    int n;
    cin >> n;
    long long matrix[n][n];
    set<long long> m;
    vector<long long> numbers;
    for (int i = 0; i < n; i++)
        for (int j = 0; j < n; j++) {
            cin >> matrix[i][j];
            m.insert(matrix[i][j]);
        }

    for (long long num : m)
        numbers.push_back(num);

    sort(numbers.begin(), numbers.end());

    for (long long num : numbers) {
        visited.clear();
        G[0].clear();
        G[201].clear();
        for (int i = 1; i <= n; i++)
            G[0].push_back(i);
        for (int i = 1; i <= 2 * n; i++)
            G[i].clear();
        for (int i = n + 1; i <= 2 * n; i++)
            G[i].push_back(201);
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++)
                if (matrix[i][j] <= num)
                    G[i + 1].push_back(n + j + 1);
        }
        if(maxFlow() == n){
            cout << num;
            break;
        }
    }
    return 0;

}

File: python/maxFlow1/main.py
n = int(input())
numbers = set()
matrix = []
for i in range(n):
        nums = input()
        matrix.append([int(x) for x in nums.split(" ")])
        numbers.update(matrix[i])
numbers = sorted(list(numbers))

def findMax():
        global numbers
        def createGraph(x):
                graph = {}
                graph[-1] = list(range(n))
                graph[200] = []
                for i in range(2 * n + 1):
                        graph[i] = []
                for i in range(n):
                        graph[n+i].append(200)
                for i in range(n):
                        for j in range(n):
                                if matrix[i][j] <= x:
                                        graph[i].append(n + j)
                return graph
        visited = []
        def dfs(graph,node):
                global visited
                visited.append(node)
                if node == 200:
                        return [200]
                for neighbor in graph[node]:
                        if neighbor not in visited:
                                visited.append(neighbor)
                                l = dfs(graph,neighbor)
                                if l != -1:
                                        return [node] + l
                return -1

        def maxFlow(graph):
                global visited
                visited = []
                l = dfs(graph,-1)

                if l == -1:
                        return 0
                for i in range(len(l) - 1):
                        a = l[i]
                        b = l[i + 1]
                        graph[a].remove(b)
                        graph[b].append(a)
                return maxFlow(graph) + 1

        for number in numbers:
                if(maxFlow(createGraph(number)) == n):
                        return number
print(findMax())